sequence/fasta_io.hpp

```diff
@@ -34,30 +48,62 @@ std::pair<Vec2D<seq_type>, std::vector<std::string>> read_fasta(const std::strin
     std::vector<seq_type> seq;
     while (std::getline(infile, line)) {
         if (line[0] == '>') {
-            seq_names.push_back(line);
+            if (seq.size()) {
+                f.sequences.push_back(std::move(seq));
+                seq.clear();
+            }
+            // Drop the leading '>'.
+            f.comments.emplace_back(line.begin() + 1, line.end());
         } else if (!line.empty()) {
             if (input_format == "fasta") {
                 for (char c : line) {
                     seq.push_back(char2int(c));
                 }
-                seqs.push_back(seq);
             } else if (input_format == "csv") {
                 std::stringstream ss(line);
                 std::string item;
                 while (std::getline(ss, item, ',')) {
                     seq.push_back(std::stoi(item, 0, 16));
                 }
-                seq_names.push_back("seq" + std::to_string(seqs.size()));
-                seqs.push_back(seq);
+                f.comments.push_back("seq" + std::to_string(f.sequences.size()));
+                f.sequences.push_back(std::move(seq));
+                seq.clear();
             } else {
                 std::cerr << "Invalid input foramt: " << input_format << std::endl;
                 exit(1);
             }
-            seq.clear();
         }
     }
-    assert(seqs.size() == seq_names.size());
-    return { seqs, seq_names };
+    if (seq.size()) {
+        f.sequences.push_back(std::move(seq));
+        seq.clear();
+    }
+    assert(f.sequences.size() == f.comments.size());
+    return f;
+}
+
+/**
+ * Reads all .fasta and .fna files in the given directory and returns them.
+ * @tparam seq_type type used for storing a character of the fasta file, typically uint8_t
+ */
+template <typename seq_type>
+std::vector<FastaFile<seq_type>> read_directory(const std::string &directory_name,
+                                                const std::string &input_format) {
+    assert(input_format == "fasta_directory");
+    if (!std::filesystem::exists(directory_name)) {
+        std::cerr << "Input directory does not exist: " << directory_name << std::endl;
+        std::exit(1);
+    }
+    std::vector<FastaFile<seq_type>> files;
+
+    for (const auto &f : std::filesystem::directory_iterator(directory_name)) {
```

Comment:
does this handle the case when directory_name is an actual file name graciously?

Reply:
It didn't. Added extra handling for this.